### PR TITLE
install servicehub json files in common7/servicehub folders

### PR DESCRIPTION
**Pull request by heejaechang**
we used to install servicehub json files in extension folder and servicehub recently added a new feature ("serviceOverride": true), where one can install its servicehub json files in Common7/ServiceHub folder (more stable) and keep existing F5 vsix experiences.

we are trying to see whether we can take advantage of this new feature so that regular VS user can get our OOP (servicehub services) from a more reliable location than extension folder. but still use private json files when dogfooding/developing.

this addresses one of our top 1 watson issue

### CODE REVIEW

**Comment by heejaechang** [comment]
@tmat I think I followed everything you said but build -pack doesn't produce new vsix file for vs setup. do I need to add new vsix name somewhere? or add new csproj as dependency to some project?

File: src/VisualStudio/Setup.ServiceHub/Roslyn.VisualStudio.Setup.ServiceHub.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<!-- Copyright (c)  Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information. -->
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net472</TargetFramework>
+    <RoslynProjectType>Custom</RoslynProjectType>
+
+    <!-- VS Insertion -->
+    <TargetVsixContainerName>Roslyn.VisualStudio.Setup.ServiceHub.vsix</TargetVsixContainerName>
+    <VisualStudioInsertionComponent>Roslyn.VisualStudio.Setup.ServiceHub</VisualStudioInsertionComponent>
```
Comment (tmat) [inline src/VisualStudio/Setup.ServiceHub/Roslyn.VisualStudio.Setup.ServiceHub.csproj]: >Roslyn.VisualStudio.Setup.ServiceHub [](start = 36, length = 36)

This should be `Microsoft.CodeAnalysis.LanguageServices` - we'll include the VSIX in our existing component

Comment (heejaechang) [reply src/VisualStudio/Setup.ServiceHub/Roslyn.VisualStudio.Setup.ServiceHub.csproj]: okay

File: src/VisualStudio/Setup/Roslyn.VisualStudio.Setup.csproj [hunk 2]
```diff
@@ -283,4 +251,11 @@
     <PackageReference Include="Microsoft.VisualStudio.Shell.Framework" Version="$(MicrosoftVisualStudioShellFrameworkVersion)" />
     <PackageReference Include="SQLitePCLRaw.bundle_green" Version="$(SQLitePCLRawbundle_greenVersion)" />
   </ItemGroup>
+  <ItemGroup>
+    <ServiceHubService Include="roslynCodeAnalysis" ClassName="Microsoft.CodeAnalysis.Remote.CodeAnalysisService" />
```
Comment (heejaechang) [inline src/VisualStudio/Setup/Roslyn.VisualStudio.Setup.csproj]: @tmat so this is duplicated 2 times. should I create props file? or put in versions.props file?

...

and this name is in vsixmanifest file. is there a way to share name here and in vsixmanifest file?

File: eng/targets/GenerateServiceHubConfigurationFiles.targets [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+}
+]]>
+        </Content>
+      </_JsonFile>
+    </ItemGroup>
+  </Target>
+
+  <Target Name="GenerateServiceHubConfigurationFiles"
+          DependsOnTargets="CalculateServiceHubConfigurationFiles"
+          AfterTargets="GetVsixSourceItems"
```
Comment (tmat) [inline eng/targets/GenerateServiceHubConfigurationFiles.targets]: >GetVsixSourceItems [](start = 24, length = 18)

I don't think this is correct. The target needs to run before `CreateVsixContainer` target runs. Looking at Microsoft.VsSDK.targets, it lokos like the best way to plug into the pipeline is via GetVsixSourceItemsDependsOn:

```xml
<PropertyGroup>
  <GetVsixSourceItemsDependsOn>$(GetVsixSourceItemsDependsOn);GenerateServiceHubConfigurationFiles</GetVsixSourceItemsDependsOn>
</PropertyGroup>
```

Comment (heejaechang) [reply eng/targets/GenerateServiceHubConfigurationFiles.targets]: okay. thank you.

File: src/Setup/DevDivVsix/ServiceHubConfig/Roslyn.VisualStudio.Setup.ServiceHub.csproj [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net472</TargetFramework>
+    <RoslynProjectType>Custom</RoslynProjectType>
+
+    <!-- VS Insertion -->
+    <TargetVsixContainerName>Roslyn.VisualStudio.Setup.ServiceHub.vsix</TargetVsixContainerName>
+    <VisualStudioInsertionComponent>Microsoft.CodeAnalysis.LanguageServices</VisualStudioInsertionComponent>
+
+    <ExtensionInstallationRoot>CommonExtensions</ExtensionInstallationRoot>
```
Comment (heejaechang) [inline src/Setup/DevDivVsix/ServiceHubConfig/Roslyn.VisualStudio.Setup.ServiceHub.csproj]: @tmat another duplicated string. what would be right way to share this with Roslyn.VS.Setup.csproj?

**Comment by heejaechang** [comment]
@tmat can you take a look?

**Comment by heejaechang** [comment]
@jinujoseph I think we want to bring this to either preview 2 or preview 3.

it has corresponding vso change (https://devdiv.visualstudio.com/DevDiv/_git/VS/pullrequest/174834?_a=files&path=%2Fsrc%2FSetupPackages%2FGroup%2FComponent%2FRoslyn%2FLanguageServices%2Fcomponent.swr) but 2 doesn't need to be inserted together. this can go in first and then vso one once this is inserted to d16.1(stg)

**Comment by jinujoseph** [comment]
Lets try to get this for preview3

File: eng/targets/GenerateServiceHubConfigurationFiles.targets [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    <ServiceHubService Include="roslynSnapshot" ClassName="Microsoft.CodeAnalysis.Remote.SnapshotService" />
+    <ServiceHubService Include="roslynRemoteSymbolSearchUpdateEngine" ClassName="Microsoft.CodeAnalysis.Remote.RemoteSymbolSearchUpdateEngine" />
+  </ItemGroup>
+
+  <PropertyGroup>
+    <GetVsixSourceItemsDependsOn>$(GetVsixSourceItemsDependsOn);GenerateServiceHubConfigurationFiles</GetVsixSourceItemsDependsOn>
+  </PropertyGroup>
+
+  <Target Name="CalculateServiceHubConfigurationFiles"
+          DependsOnTargets="GetAssemblyVersion">
```
Comment (tmat) [inline eng/targets/GenerateServiceHubConfigurationFiles.targets]: >DependsOnTargets="GetAssemblyVersion" [](start = 10, length = 37)

We don't need assembly version anymore.

File: src/Setup/DevDivVsix/ServiceHubConfig/Roslyn.VisualStudio.Setup.ServiceHub.csproj [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+<Project Sdk="Microsoft.NET.Sdk">
+  <PropertyGroup>
+    <TargetFramework>net472</TargetFramework>
+    <RoslynProjectType>Custom</RoslynProjectType>
+
+    <!-- VS Insertion -->
+    <TargetVsixContainerName>Roslyn.VisualStudio.Setup.ServiceHub.vsix</TargetVsixContainerName>
+    <VisualStudioInsertionComponent>Microsoft.CodeAnalysis.LanguageServices</VisualStudioInsertionComponent>
+
+    <!-- ServiceHub AssemblyPath -->
```
Comment (tmat) [inline src/Setup/DevDivVsix/ServiceHubConfig/Roslyn.VisualStudio.Setup.ServiceHub.csproj]: >ServiceHub AssemblyPath [](start = 9, length = 23)

Document what this path is relative to.

File: src/VisualStudio/Setup/Roslyn.VisualStudio.Setup.csproj [hunk 7]
```diff
@@ -12,11 +12,14 @@
...
     <IncludeDebugSymbolsInLocalVSIXDeployment>false</IncludeDebugSymbolsInLocalVSIXDeployment>
-    <ExtensionInstallationRoot>CommonExtensions</ExtensionInstallationRoot>
-    <ExtensionInstallationFolder>Microsoft\ManagedLanguages\VBCSharp\LanguageServices</ExtensionInstallationFolder>
+    <ExtensionInstallationRoot>$(CommonExtensionInstallationRoot)</ExtensionInstallationRoot>
+    <ExtensionInstallationFolder>$(LanguageServicesExtensionInstallationFolder)</ExtensionInstallationFolder>
 
     <!-- VS Insertion -->
     <VisualStudioInsertionComponent>Microsoft.CodeAnalysis.LanguageServices</VisualStudioInsertionComponent>
+  
+    <!-- ServiceHub AssemblyPath -->
```
Comment (tmat) [inline src/VisualStudio/Setup/Roslyn.VisualStudio.Setup.csproj]: >ServiceHub AssemblyPath [](start = 9, length = 23)

Document what this path is relative to and what it is. E.g. "Path to a directory that contains servicehub.service.json files, relative to ..."

**Pull request merged**